### PR TITLE
Translate advanced_composition code from dp_accounting library to Kotlin

### PR DESCRIPTION
**Pull request by matthewclegg**
See the following:

https://github.com/google/differential-privacy/blob/main/python/dp_accounting/accountant.py

---
This change is [Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/467)

### CODE REVIEW

File: src/main/kotlin/org/wfanet/measurement/eventdataprovider/privacybudgetmanagement/AdvancedComposition.kt [hunk 1]
```diff
@@ -0,0 +1,93 @@
...
+   * @return The number of distinct ways to draw k items from a set of size n. Alternatively, the
+   * coefficient of x^k in the expansion of (1 + x)^n.
+   */
+  fun coeff(n: Int, k: Int): Float {
+    if ((n < 0) || (k < 0) || (n < k)) {
+      return 0.0f
+    } else if ((k == 0) || (n == k)) {
+      return 1.0f
+    } else if (memoizedCoeffs.containsKey(Pair(n, k))) {
+      return memoizedCoeffs.get(Pair(n, k)) ?: 0.0f
```
Comment (pasin30055) [inline src/main/kotlin/org/wfanet/measurement/eventdataprovider/privacybudgetmanagement/AdvancedComposition.kt:37]: Optional: we can also let `k = n - k` in the case where k > n/2 to reduce the running time / memory further.

**Review by pasin30055 — APPROVED** [review]
I don't know kotlin syntax / conventions well, but from what I can understand, this looks good to me. Thanks!

**Review by stevenwarejones — COMMENTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/467)* status: 0 of 3 files reviewed, 2 unresolved discussions (waiting on @matthewclegg and @uakyol)

___
*[src/main/kotlin/org/wfanet/measurement/eventdataprovider/privacybudgetmanagement/AdvancedComposition.kt, line 21 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/467#-MxaSP3R1G_3_hh6Ora_:-MxaSP3R1G_3_hh6Oraa:b1wdomn) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/34202451aff4304800bd433d2460a14dae47a680/src/main/kotlin/org/wfanet/measurement/eventdataprovider/privacybudgetmanagement/AdvancedComposition.kt#L21)):*
> ```Kotlin
> /** Memoized computation of binomial coefficients. */
> class Binom {
>   var memoizedCoeffs = HashMap<Pair<Int, Int>, Float>()
> ```

do we need a ConcurrentHashMap?

**Review by matthewclegg — COMMENTED** [review]
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/467)* status: 0 of 3 files reviewed, 2 unresolved discussions (waiting on @pasin30055, @stevenwarejones, and @uakyol)

___
*[src/main/kotlin/org/wfanet/measurement/eventdataprovider/privacybudgetmanagement/AdvancedComposition.kt, line 21 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/467#-MxaSP3R1G_3_hh6Ora_:-Mxl8iTt0siZXLqCiy4P:b-hbnax6) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/34202451aff4304800bd433d2460a14dae47a680/src/main/kotlin/org/wfanet/measurement/eventdataprovider/privacybudgetmanagement/AdvancedComposition.kt#L21)):*
<details><summary><i>Previously, stevenwarejones (Steven Ware Jones) wrote…</i></summary><blockquote>

do we need a ConcurrentHashMap?
</blockquote></details>

The memoizedCoeffs field is specific to each object instance, and a new object instance is created in each call to totalPrivaceBudgetUsageUnderAdvancedComposition, which only lives for the life of that function call.  There is no opportunity for memoizedCoeffs to be accessed by two different threads.  So, there would not be any benefit to using a ConcurrentHashMap.  Since a ConcurrentHashMap has higher overhead than a HashMap, it is better to use a HashMap.
___
*[src/main/kotlin/org/wfanet/measurement/eventdataprovider/privacybudgetmanagement/AdvancedComposition.kt, line 37 at r1](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/467#-MxT-EkSxBdGOai1LVy2-r1-37:-Mxl9hlO2Zshkp-SS1ds:b-896fix) ([raw file](https://github.com/world-federation-of-advertisers/cross-media-measurement/blob/34202451aff4304800bd433d2460a14dae47a680/src/main/kotlin/org/wfanet/measurement/eventdataprovider/privacybudgetmanagement/AdvancedComposition.kt#L37)):*
<details><summary><i>Previously, pasin30055 (Pasin Manurangsi) wrote…</i></summary><blockquote>

Optional: we can also let `k = n - k` in the case where k > n/2 to reduce the running time / memory further.
</blockquote></details>

Done.

**Review by stevenwarejones — APPROVED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/467)* status: 0 of 3 files reviewed, 1 unresolved discussion (waiting on @pasin30055 and @uakyol)

**Review by matthewclegg — COMMENTED** [review]
:lgtm:

> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/467#-:-My50JVlAqIugvI9mQbX:bnfp4nl)* status: 0 of 3 files reviewed, 1 unresolved discussion (waiting on @pasin30055 and @uakyol)

**Review by matthewclegg — COMMENTED** [review]
> Dismissed @pasin30055 from a discussion.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/467)* status: 0 of 4 files reviewed, all discussions resolved (waiting on @pasin30055 and @stevenwarejones)

**Review by matthewclegg — COMMENTED** [review]
> Reviewed 2 of 2 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/world-federation-of-advertisers/cross-media-measurement/467)* status: 2 of 4 files reviewed, all discussions resolved (waiting on @pasin30055 and @stevenwarejones)

**Pull request merged**